ad9959/src/lib.rs

```diff
@@ -511,6 +491,72 @@ impl<I: Interface> Ad9959<I> {
     }
 }
 
+/// Validate the internal system clock configuration of the chip.
+///
+/// Arguments:
+/// * `reference_clock_frequency` - The reference clock frequency provided to the AD9959 core.
+/// * `multiplier` - The frequency multiplier of the system clock. Must be 1 or 4-20.
+///
+/// Returns:
+/// The system clock frequency to be configured.
+pub fn validate_clocking(
+    reference_clock_frequency: f32,
+    multiplier: u8,
+) -> Result<f32, Error> {
+    if multiplier != 1 && !(4..=20).contains(&multiplier)
+        || (multiplier != 1 && reference_clock_frequency < 10e6)
+        || reference_clock_frequency < 1e6
+    {
+        return Err(Error::Bounds);
+    }
+
+    let frequency = multiplier as f32 * reference_clock_frequency;
+    if !(255e6..=500e6).contains(&frequency)
+        && !(100e6..=160e6).contains(&frequency)
+        && (multiplier != 1 || !(0.0..100e6).contains(&frequency))
+    {
+        return Err(Error::Frequency);
+    }
+
+    Ok(frequency)
+}
+
+pub fn frequency_to_ftw(
+    dds_frequency: f32,
+    system_clock_frequency: f32,
+) -> Result<u32, Error> {
+    if !(0.0..=(system_clock_frequency / 2.0)).contains(&dds_frequency) {
+        return Err(Error::Bounds);
+    }
+    // The function for channel frequency is `f_out = FTW * f_s / 2^32`, where FTW is the
+    // frequency tuning word and f_s is the system clock rate.
+    Ok(((dds_frequency / system_clock_frequency) * (1u64 << 32) as f32) as u32)
+}
+
+pub fn phase_to_pow(phase_turns: f32) -> Result<u16, Error> {
+    Ok((phase_turns * (1 << 14) as f32) as u16 & ((1 << 14) - 1))
```

Comment:
I think this conversion could definitely use some documentation. It feels like we could also just simplify it. Why is (1<< 14) - 1 better than just `0x3FFF`?